clippy_utils/src/higher.rs

```diff
@@ -416,56 +419,169 @@ pub fn binop(op: hir::BinOpKind) -> ast::BinOpKind {
+    /// second argument of the asset for case `assert_eq!`,
+    /// `assert_ne!` etc ... Eg var_2 in `debug_assert_eq!(x, var_2,..)`
+    pub second_assert_argument: Option<&'tcx Expr<'tcx>>,
+    /// The format argument passed at the end of the macro
+    pub format_arg: Option<FormatArgsExpn<'tcx>>,
+}
+
+impl<'tcx> AssertExpn<'tcx> {
+    /// Extract args from an assert-like macro.
+    /// Currently working with:
+    /// - `assert!`, `assert_eq!` and `assert_ne!`
+    /// - `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!`
+    /// For example:
+    /// `assert!(expr)` will return `Some(AssertExpn { first_assert_argument: expr,
+    /// second_assert_argument: None, format_arg:None })` `debug_assert_eq!(a, b)` will return
+    /// `Some(AssertExpn { first_assert_argument: a, second_assert_argument: Some(b),
+    /// format_arg:None })`
+    pub fn parse(e: &'tcx Expr<'tcx>) -> Option<Self> {
+        if let ExprKind::Block(block, _) = e.kind {
+            if block.stmts.len() == 1 {
+                if let StmtKind::Semi(matchexpr) = block.stmts.get(0)?.kind {
+                    // macros with unique arg: `{debug_}assert!` (e.g., `debug_assert!(some_condition)`)
+                    if_chain! {
+                        if let Some(If { cond, then, .. }) = If::hir(matchexpr);
+                        if let ExprKind::Unary(UnOp::Not, condition) = cond.kind;
+                        then {
+                            if_chain! {
+                                if let ExprKind::Block(block, _) = then.kind;
+                                if let [statement, ..] = block.stmts;
+                                if let StmtKind::Expr(call_assert_failed)
+                                    | StmtKind::Semi(call_assert_failed) = statement.kind;
+                                if let ExprKind::Call(_, args_assert_failed) = call_assert_failed.kind;
+                                if !args_assert_failed.is_empty();
+                                if let ExprKind::Call(_, [arg, ..]) =  args_assert_failed[0].kind;
+                                if let Some(format_arg_expn) = FormatArgsExpn::parse(arg);
+                                then {
+                                    return Some(Self {
+                                        first_assert_argument: condition,
+                                        second_assert_argument: None,
+                                        format_arg: Some(format_arg_expn), // FIXME actually parse the aguments
+                                    });
+                                }
+                                else{
+                                    return Some(Self {
+                                        first_assert_argument: condition,
+                                        second_assert_argument: None,
+                                        format_arg: None,
+                                    });
+                                }
```

Comment:
Try to avoid `else` in `if_chain!` here since it adds an else clause to every `if` in the expansion. You should only need to "fork" at a single point.